src/pages/product-details/comments/comments.js

```diff
@@ -1,74 +1,258 @@
-import React, { useState } from 'react';
-import { useSelector } from 'react-redux';
+import React, { useState, useEffect, useMemo } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
 
 import Rating from '@material-ui/lab/Rating';
 import TextField from '@material-ui/core/TextField';
+import Tooltip from '@material-ui/core/Tooltip';
 import Button from '@material-ui/core/Button';
+import useValidation from '../../../utils/use-validation';
 import useStyles from './comments.styles';
 
+import { addComment } from '../../../redux/products/products.actions';
 import CommentsItem from './comments-item';
+import SnackbarItem from '../../../containers/snackbar';
+import { Loader } from '../../../components/loader/loader';
 
-import { INPUT_VARIANT } from '../../../configs';
+import {
+  COMMENT_DATA,
+  TEXT,
+  formRegExp,
+  errorMessages
+} from '../../../configs';
 import { COMMENTS } from '../../../translations/product-details.translations';
 
 const Comments = () => {
   const styles = useStyles();
-  const { language, comments } = useSelector(({ Language, Products }) => ({
+  const dispatch = useDispatch();
+
+  const {
+    commentsLoading,
+    language,
+    productId,
+    comments,
+    userData
+  } = useSelector(({ Products, Language, User }) => ({
+    commentsLoading: Products.commentsLoading,
+    productId: Products.product._id,
+    comments: Products.product.comments,
     language: Language.language,
-    comments: Products.product.comments
+    userData: User.userData
   }));
 
+  const {
+    firstNameValidated,
+    emailValidated,
+    textValidated,
+    allFieldsValidated,
+    shouldValidate,
+    comment,
+    setFirstNameValidated,
+    setEmailValidated,
+    setTextValidated,
+    setAllFieldsValidated,
+    setShouldValidate,
+    setComment,
+    clearFields,
+    filterText,
+    validateField
+  } = useValidation();
+
   const [rate, setRate] = useState(0);
 
+  const { _id, email: userEmail, firstName: userName, purchasedProduct } =
+    userData || {};
```

Comment:
We should not try any destructuring if no `userData` is found so `userData || {}` is pretty useless, isn't it?

Reply:
`userData ` for convenience has value of `null `in redux, so i need this condition

Reply:
I don't think that you should perform this operation at all if `userData` is not provided